### ISSUE TITLE
[BUG] SAM CLI deployed lambda returns InvalidEntryPoint error

### ISSUE THREAD

**Issue by jkruse14** [aws-cloudformation/cloudformation-guard#472]
**Describe the bug**
When deploying with SAM CLI, I am getting `Runtime.InvalidEntrypoint`

Note: I want to name the function, so I supply my own template when running in a GitHub action. I follow the steps for the SAM CLI deployment, but before hand I copy in my own template.yaml with the following content:

```
AWSTemplateFormatVersion: '2010-09-09'
Transform:
  - AWS::Serverless-2016-10-31
  - AWS::LanguageExtensions
Description: Lambda function to run cfn-guard

Parameters:
  Environment:
    Type: String
    Description: Deployment environment   #Look into reducing to the only Env we will be deploying from.
    AllowedValues:
      - cicd-test
      - dev
  CfnGuardVersion:
    Type: String
    Description: Version of cfn-guard deployed

Mappings:
  EnvironmentSpecific:
    cicd-test:
      subnets:
        - subnet-xxx
        - subnet-yyy
    dev:
      subnets:
        - subnet-xxx
        - subnet-yyy
        - subnet-zzz
Resources:
  
  # Name must be CloudFormationGuardLambda
  CloudFormationGuardLambda:
    Type: AWS::Serverless::Function
    Properties:
      Description: Lambda function to run cfn-guard rules
      CodeUri: ..
      FunctionName: cfn-guard
      PackageType: Zip
      Handler: guard.handler
      Runtime: provided.al2
      Environment:
        Variables:
          RUST_BACKTRACE: "1"
      Tracing: Active
      PropagateTags: true
      Tags:
        cfn-guard:version: !Ref CfnGuardVersion
      VpcConfig:
        SubnetIds: !FindInMap [ EnvironmentSpecific, !Ref Environment, subnets ]
        SecurityGroupIds:
          - !ImportValue my-sg
    Metadata:
      BuildMethod: makefile

Outputs:
  CfnGuardFunction:
    Description: CloudFormation Guard Lambda Function
    Value: !Ref CloudFormationGuardLambda
    Export:
      Name: !Sub ${AWS::StackName}-CfnGuardFunction
```

The primary differences:
- FunctionName
- VpcConfig (just to confirm, I've redacted the subnet IDs)
- LoggingConfig
- Adds and Export

Deployment script:

```
        cp ./cloudformation/samconfig.toml ./cloudformation-guard/guard-lambda/samconfig.toml
        echo "Copied samconfig.toml"
        # The original template does not specify a FunctionName which is needed to simplify our hook
        # This also allows for the setting of VpcConfig and other properties
        cp ./cloudformation/template.yaml ./cloudformation-guard/guard-lambda/template.yaml
        echo "Copied template.yaml"
        cd ./cloudformation-guard/guard-lambda
        pwd
        echo "Changed directory to guard-lambda"
        echo "Starting build"
        sam build --use-container --config-env github
        ls
        echo "Build complete"
        echo "Deploying cfn-guard"
        sam package --config-env github \
          --s3-bucket ${{ vars.CLOUDFORMATION_BUCKET }} \
          --s3-prefix db-aws-cloudformation-guard-fn
        sam deploy --config-env github \
          --s3-bucket ${{ vars.CLOUDFORMATION_BUCKET }} \
          --no-fail-on-empty-changeset \
          --role-arn ${{ secrets.CLOUDFORMATION_ROLE }} \
          --stack-name db-aws-cloudformation-guard-fn\
          # --parameter-overrides "Environment=${{inputs.environment}} CfnGuardVersion=${{ steps.get-version.outputs.cfn-guard-tag }}" \
```

If I comment out the lines to copy my template file over the provided, I get the same result, same error message (see below). For completeness, I also removed my `sam package` command, which yielded the same result.

Why I want to name my function: I want to use this function in a CloudFormation Hook to centralize validation rather than force all teams to update all CFN related workflows with a GitHub Action. Having a consistent function makes this easy.

**To Reproduce**
Deploy with template provided above and provided script then invoke the function

Error received in CloudWatch logs:

```
2024-02-24T12:07:43.549-06:00
RequestId: 6cdef61b-71a6-447f-95e6-921134cacf17 Error: Couldn't find valid bootstrap(s): [/var/task/bootstrap /opt/bootstrap]
Runtime.InvalidEntrypoint
```

**Expected behavior**
I *believe* our cfn templates are functionally equivalent, so I would expect the function to execute properly (i.e. confirm or deny the resource meets the rule's standards)

**Operating System:**
default lambda OS (AL2 based on `Runtime`?)

**OS Version**
default lambda OS (AL2 based on `Runtime`?)

**Comment by jkruse14** [comment]
I found the docs for [building a custom runtime](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-custom.html). It has the following:

> Entrypoint
> 
> A custom runtime's entry point is an executable file named bootstrap. The bootstrap file can be the runtime, or it can invoke another file that creates the runtime. If the root of your deployment package doesn't contain a file named bootstrap, Lambda looks for the file in the function's layers. If the bootstrap file doesn't exist or isn't executable, your function returns a Runtime.InvalidEntrypoint error upon invocation.

This suggests my issue is the `bootstrap` file cannot be found or is not executable.

I'm wondering if add `chmod 755` is needed in the Make file?

It feels like there is something missing in the SAM CLI steps, but I'm not familiar with custom runtimes.

**Comment by jkruse14** [comment]
I downloaded the archive sent to CloudFormation

`guard-lambda/.aws_sam/`:

`build toml`:
```
# This file is auto generated by SAM CLI build command

[function_build_definitions.0d073aed-090f-4c83-84e2-c81a4f2c1d63]
codeuri = "/home/runner/work/db-aws-cloudformation-guard/db-aws-cloudformation-guard/cloudformation-guard"
runtime = "provided.al2"
architecture = "x86_64"
handler = "guard.handler"
manifest_hash = ""
packagetype = "Zip"
functions = ["CloudFormationGuardLambda"]

[function_build_definitions.0d073aed-090f-4c83-84e2-c81a4f2c1d63.metadata]
BuildMethod = "makefile"

[layer_build_definitions]
```

`build/template.yaml` - my template posted above

`build/CloudFormationGuardLambda/`: bootstrap archive file

Within the archive is see:
<img width="592" alt="image" src="https://github.com/aws-cloudformation/cloudformation-guard/assets/16331726/13f1f6fc-7993-4cbb-84a2-15eb43360e34">

I'm guessing this issue is I should not have the entire repo packaged and uploaded to S3 and only the `.aws_sam` directory?

**Comment by jkruse14** [comment]
I found my issue - it was two fold.

First, I was not familiar with `sam build` vs. `sam package` as I've only really had need for `sam package`.  After digging deeper on this, I found `build` creates the `.aws_sam` directory and `deploy` will upload this ✔️ 

Second, I was using a `samconfig.toml` file which specified the template. This needed to be removed because it was overriding the template in `.aws_config` ✔️ 

With these issues fixed, I have a working lambda function running cfn-guard! 🎉
